flags/mod.ts

```diff
@@ -6,17 +6,204 @@
+  & C
+  & S
+) ? true
+  : false;
+
+/**
+ * Creates a record with all available flags with the corresponding type and
+ * default type.
+ */
+type Values<
+  B extends BooleanType,
+  S extends StringType,
+  C extends CollectType,
+  D extends Record<string, unknown> | undefined,
+  A extends Aliases | undefined,
+> // deno-lint-ignore no-explicit-any
+ = NoTypes<B, S, C> extends true ? Record<string, any>
+  : 
+    & Record<string, unknown>
+    & AddAliases<
+      SpreadDefaults<
+        & CollectValues<S, string, C>
+        & RecursiveRequired<CollectValues<B, boolean, C>>
+        & UnknownCollactable<B, S, C>,
+        DedotRecord<D>
+      >,
+      A
+    >;
+
+type Aliases<T = string, V extends string = string> = Partial<
+  Record<Extract<T, string>, V | ReadonlyArray<V>>
+>;
+
+type AddAliases<
+  T,
+  A extends Aliases | undefined,
+> = { [K in keyof T as AliasName<K, A>]: T[K] };
+
+type AliasName<
+  K,
+  A extends Aliases | undefined,
+> = K extends keyof A
+  ? string extends A[K] ? K : A[K] extends string ? K | A[K] : K
+  : K;
+
+/**
+ * Spreads all default values of Record `D` into Record `A`
+ * and makes default values required.
+ *
+ * **Example:**
+ * `SpreadValues<{ foo?: boolean, bar?: number }, { foo: number }>`
+ *
+ * **Result:** `{ foo: boolan | number, bar?: number }`
+ */
+type SpreadDefaults<A, D> = D extends undefined ? A
+  : A extends Record<string, unknown> ? 
+    & Omit<A, keyof D>
+    & {
+      [K in keyof D]: K extends keyof A
+        ? (A[K] & D[K] | D[K]) extends Record<string, unknown>
+          ? NonNullable<SpreadDefaults<A[K], D[K]>>
+        : D[K] | NonNullable<A[K]>
+        : unknown;
+    }
+  : never;
+
+/**
+ * Defines the Record for the `default` option to add
+ * auto suggestion support for IDE's.
+ */
+type Defaults<B extends BooleanType, S extends StringType> = Id<
+  UnionToIntersection<
+    & Record<string, unknown>
+    // Dedotted auto suggestions: { foo: { bar: unknown } }
+    & MapTypes<S, unknown>
+    & MapTypes<B, unknown>
+    // Flat auto suggestions: { "foo.bar": unknown }
+    & MapDefaults<B>
+    & MapDefaults<S>
+  >
+>;
+
+type MapDefaults<T extends ArgType> = Partial<
+  Record<T extends string ? T : string, unknown>
+>;
+
+type RecursiveRequired<T> = T extends Record<string, unknown> ? {
+  [K in keyof T]-?: RecursiveRequired<T[K]>;
+}
+  : T;
+
+/** Same as `MapTypes` but also supports collectable options. */
+type CollectValues<T extends ArgType, V, C extends CollectType> =
+  UnionToIntersection<
+    C extends string ? 
+      & MapTypes<Exclude<T, C>, V>
+      & (T extends undefined ? Record<never, never> : RecursiveRequired<
+        MapTypes<Extract<C, T>, Array<V>>
+      >)
+      : MapTypes<T, V>
+  >;
+
+/** Same as `Record` but also supports dotted and negatable options. */
+type MapTypes<T extends ArgType, V> = undefined extends T ? Record<never, never>
+  : T extends `no-${infer Name}` ? MapTypes<Name, V | false>
```

Comment:
I think the last thing that is missing is the handling of negatable options. How do we want to proceed with them? do we want to make them configurable by prefixing them with `no-` or with a separate option?

Reply:
@c4spar 
Sorry for the delay in response.

> do we want to make them configurable by prefixing them with no- or with a separate option?

I don't think we should change it optional. That feels like an unnecessary breaking change to me. In my view `--no-foo` shorthand is always useful for boolean args. So I suggest we just keep the existing behavior

For negation of non-boolean args, I can't think of reasonable usages of it. So I'd suggest we should leave it undefined behavior and let's ignore it in typings. What do you think?

Reply:
I still think we should keep support for negation of non boolean args.
A good example is `deno run --no-config` 😉 It's a string arg with a default behavior that can be disabled.

But if we don't make it configurable than all string and collectable args can be also `false`, which is a bit annoying.

Reply:
Ok. `--no-config` looks like a useful example.

> But if we don't make it configurable than all string and collectable args can be also false, which is a bit annoying.

I agree with this. What API do you suggest for configurable negatable args?

Is the following example what you suggest for configurable negatable args? (I think this is not implemented though)
```
    boolean: ["foo", "no-bar", "no-dotted.tick", "dotted.tock"],
    string: ["beep", "no-boop", "no-dotted.zig", "dotted.zag"],
```

Reply:
> Is the following example what you suggest for configurable negatable args? (I think this is not implemented though)

This would be the easiest solution and the typings for this are already implemented in this PR. I have also implemented the logic but only locally, i can open another PR for this.

Another solution would be to add a new `negate` option, same like the `collect` option. But the typings would be more complex for this solution.

Reply:
@c4spar 
> This would be the easiest solution and the typings for this are already implemented in this PR. I have also implemented the logic but only locally, i can open another PR for this.

Thank you for your efforts on this, but this syntax (`boolean: ["no-bar"]`) for indicating negatable arg looks confusing/too implicit to me.

> Another solution would be to add a new negate option, same like the collect option. But the typings would be more complex for this solution.

This solution looks much better to me as it's more explicit and probably easier to document. Also this feels consistent with `collect` option for configuring arg behaviors.

Reply:
ok i got the types for the `negatable` option working now 😉